src/app/cli/src/init/transaction_snark_profiler.ml

```diff
@@ -154,61 +198,111 @@ let profile (module T : Transaction_snark.S) sparse_ledger0
-        in
-        let span, proof =
-          time (fun () ->
-              Async.Thread_safe.block_on_async_exn (fun () ->
-                  T.of_non_parties_transaction
-                    ~statement:
-                      { sok_digest = Sok_message.Digest.default
-                      ; source =
-                          { ledger = Sparse_ledger.merkle_root sparse_ledger
-                          ; pending_coinbase_stack = coinbase_stack_source
-                          ; local_state = Mina_state.Local_state.empty ()
-                          }
-                      ; target =
-                          { ledger = Sparse_ledger.merkle_root sparse_ledger'
-                          ; pending_coinbase_stack = coinbase_stack_target
-                          ; local_state = Mina_state.Local_state.empty ()
-                          }
-                      ; supply_increase
-                      ; fee_excess =
-                          Transaction.fee_excess (Transaction.forget t)
-                          |> Or_error.ok_exn
-                      }
-                    ~init_stack:coinbase_stack_source
-                    { Transaction_protocol_state.Poly.transaction = t
-                    ; block_data = Lazy.force state_body
-                    }
-                    (unstage (Sparse_ledger.handler sparse_ledger)) ) )
+        let tm0 = Core.Unix.gettimeofday () in
+        let%map proof =
+          T.of_non_parties_transaction
+            ~statement:
+              { sok_digest = Sok_message.Digest.default
+              ; source =
+                  { ledger = Sparse_ledger.merkle_root sparse_ledger
+                  ; pending_coinbase_stack = coinbase_stack_source
+                  ; local_state = Mina_state.Local_state.empty ()
+                  }
+              ; target =
+                  { ledger = Sparse_ledger.merkle_root sparse_ledger'
+                  ; pending_coinbase_stack = coinbase_stack_target
+                  ; local_state = Mina_state.Local_state.empty ()
+                  }
+              ; supply_increase =
+                  (let magnitude =
+                     Transaction.expected_supply_increase t |> Or_error.ok_exn
+                   in
+                   let sgn = Sgn.Pos in
+                   Currency.Amount.Signed.create ~magnitude ~sgn )
+              ; fee_excess =
+                  Transaction.fee_excess (Transaction.forget t)
+                  |> Or_error.ok_exn
+              }
+            ~init_stack:coinbase_stack_source
+            { Transaction_protocol_state.Poly.transaction = t
+            ; block_data = Lazy.force state_body
+            }
+            (unstage (Sparse_ledger.handler sparse_ledger))
         in
+        let tm1 = Core.Unix.gettimeofday () in
+        let span = Time.Span.of_sec (tm1 -. tm0) in
         ( (Time.Span.max span max_span, sparse_ledger', coinbase_stack_target)
-        , proof ) )
+        , proofs @ [ proof ] ) )
```

Comment:
Accumulate a list rather than appending? As written, this will do `O(n^2)` allocations instead of `O(n)`.

Reply:
Right. The original code used `List.fold_map`, which is not available in `Deferred.List`.